e3sm_diags/run.py

```diff
@@ -18,79 +20,257 @@ class Run:
-            msg = "You must pass in a list of parameter objects."
-            raise RuntimeError(msg)
+        self._validate_parameters(parameters)
+
+        # FIXME: This line produces some unintended side-effects. For example,
+        # let's say we have two objects: 1. CoreParameter, 2. ZonalMean2DParameter.
+        # If object 1 has `plevs=[200]`, this value will get copied to object 2.
+        # Object 2 has a check to make sure plevs has more than 1 value. This
+        # breaks the diagnostic run as a result. The workaround is to loop
+        # over `run_diags()` function and run one parameter at a time.
+        self._add_parent_attrs_to_children(parameters)
 
-        # For each of the passed in parameters, we can only have one of
-        # each type.
-        types = set([p.__class__ for p in parameters])
-        if len(types) != len(parameters):
-            msg = "You passed in two or more parameters of the same type."
-            raise RuntimeError(msg)
+        if not debug:
+            run_params = self._get_cfg_parameters(parameters)
+        else:
+            run_params = self._get_debug_parameters(parameters)
 
-        self._add_parent_attrs_to_children(parameters)
+        self.parser.check_values_of_params(run_params)
 
-        final_params = []
+        return run_params
 
-        for set_name in self.sets_to_run:
-            other_params = self._get_other_diags(parameters[0].run_type)
+    def _validate_parameters(self, parameters: List[CoreParameter]):
+        if parameters is None or not isinstance(parameters, list):
+            raise RuntimeError("You must pass in a list of parameter objects.")
+
+        param_types_list = [
+            p.__class__ for p in parameters if p.__class__ != CoreParameter
+        ]
+        param_types_set = set(param_types_list)
+
+        if len(param_types_set) != len(param_types_list):
+            raise RuntimeError(
+                "You passed in two or more non-CoreParameter objects of the same type."
+            )
 
+    def _get_cfg_parameters(
+        self, parameters: List[CoreParameter]
+    ) -> List[CoreParameter]:
+        """Get the run parameters using all sets and a .cfg file if passed.
+
+        Parameters
+        ----------
+        parameters : List[CoreParameter]
+            A list of parameter objects.
+
+        Returns
+        -------
+        List[CoreParameter]
+            A list of parameter objects including ones defined in a .cfg file.
+            Any non-CoreParameter objects will be replaced by a sub-class
+            based on the set (``SETS_TO_PARAMETERS``).
+        """
+        run_params = []
+
+        # Get parameters from user-defined .cfg file or default diags .cfg
+        # file.
+        if self.has_cfg_params:
+            cfg_params = self._get_diags_from_cfg_file()
+        else:
+            run_type = parameters[0].run_type
+            cfg_params = self._get_default_diags_from_cfg_file(run_type)
```

Comment:
This block of code splits up how `.cfg` parameters are retrieved, either by an explicitly defined .cfg file (`-d`) or default diagnostics.